birdset/datamodule/embeddings_datamodule.py

```diff
@@ -0,0 +1,293 @@
+                    testval_count[label] += 1    
+
+            
+            # Create and print table to show class distribution
+            headers = ["Class", "#Train-Samples", "#Test,Valid-Samples"]
+            rows = []
+            
+            for class_id in selected_samples.keys():
+                rows.append([self.id_to_label[class_id], train_count[class_id], testval_count[class_id]])
+            
+            print(tabulate(rows, headers, tablefmt="rounded_grid"))
+            
+            # Flatten the selected samples into a single list
+            selected_samples = [sample for samples in selected_samples.values() for sample in samples]
+
+            # Split the selected samples into training, validation, and testing sets
+
+            if self.val_batches == 0:
+                train_data = selected_samples
+                test_data = rest_samples
+                val_data = Dataset.from_dict({})
+            
+            else:    
+                num_samples = len(rest_samples)
+                num_test_samples = int(self.test_ratio * num_samples)
+
+                train_data = selected_samples
+                test_data = rest_samples[:num_test_samples]
+                val_data = rest_samples[num_test_samples:]
+                val_data = Dataset.from_dict({key: [sample[key] for sample in val_data] for key in val_data[0]}) #! Use first test sample as val cant be empty
+            
+            train_data = Dataset.from_dict({key: [sample[key] for sample in train_data] for key in train_data[0]})
+            test_data = Dataset.from_dict({key: [sample[key] for sample in test_data] for key in test_data[0]})
+
+            # Combine into a DatasetDict
+            dataset = DatasetDict({
+                'train': train_data,
+                'valid': val_data,
+                'test': test_data
+            })
+        else:
+            if self.val_batches == 0:
+                dataset['test'] = concatenate_datasets([dataset['valid'], dataset['test']])
+                dataset['valid'] = Dataset.from_dict({}) #! So that no split is created in the dataloader
+                
+            if self.low_train:
+                del dataset['train']
+                dataset['train'] = dataset['train_low']
+
+            del dataset['train_low']    
+                
+            
+            
+        return dataset
+
+    def _compute_embeddings(self, dataset):
+        """
+        Compute Embeddings for the entire dataset and store them in a new DatasetDict.
+        """
+        # Create a new DatasetDict to store the embeddings
+        embeddings_dataset = DatasetDict()
+        
+        #! For some reason the .map() from Datasets caused errors
+        # Iterate over each split in the dataset
+        for split in dataset.keys():
+            print(f">> Extracting Embeddings for {split} Split")
+            # Get the current split data
+            split_data = dataset[split]
```

Comment:
@XgamerTV Does this make a deep copy or only a copy by reference?

Reply:
I think it's a copy by reference but the hf dataset objects are immutable by default as functions like map should be used (which didn't work for me). I tried to chagne the code but had no success.

Reply:
Ah okay, thanks for explaining!

Reply:
Ah I actually managed to use the map function now!